### PR TITLE
Fix #error:Expected url scheme

### PR DESCRIPTION
**Pull request by dsvoronin**

### CODE REVIEW

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/transport/NoOpTransport.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    private fun sendContentInternal(
+        comment: String,
+        type: Entry.File.Type
+    ): FutureValue<Entry.File> {
+        return FutureValue.create(
+            Entry.File(
+                comment = comment,
+                fileAddress = FileAddress.Error(RuntimeException("File not available: NoOpTransport chosen")),
```
Comment (eugene-krivobokov) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/transport/NoOpTransport.kt]:
```suggestion
                fileAddress = FileAddress.Error(RuntimeException("File is not available: NoOpTransport chosen")),
```

File: subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/HttpRemoteStorage.kt [hunk 2]
```diff
@@ -1,178 +1,94 @@
...
+            )
+        )
+    }
 
-        val timestamp = timeProvider.nowInSeconds()
+    private fun ContentType.toMediaType(): MediaType = when (this) {
+        ContentType.PNG -> "image/png"
+        ContentType.MP4 -> "video/mp4"
+        ContentType.TXT -> "text/plain"
+        ContentType.HTML -> "text/html"
```
Comment (eugene-krivobokov) [inline subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/HttpRemoteStorage.kt]: We force clients of `ContentType` to duplicate such mappings.
Let's incorporate this knowledge into `ContentType` itself.

File: subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt [hunk 3]
```diff
@@ -91,6 +93,21 @@ sealed class Result<T> {
         }
     }
 
+    inline fun <R> combine(other: Result<T>, func: (T, T) -> R): Result<R> {
```
Comment (eugene-krivobokov) [inline subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt:96]:
```suggestion
    inline fun <R> combine(other: Result<T>, onSuccess: (T, T) -> R): Result<R> {
```
?

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/common/okhttp/src/main/kotlin/com/avito/http/HttpUrl.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+
+public fun String.toHttpUrlResult(): Result<HttpUrl> {
+    return try {
+        Result.Success(toHttpUrl())
+    } catch (e: IllegalArgumentException) {
+        Result.Failure(IllegalArgumentException("Failed to parse httpUrl: $this", e))
+    }
+}
```
Comment (eugene-krivobokov) [inline subprojects/common/okhttp/src/main/kotlin/com/avito/http/HttpUrl.kt:13]: I'm thinking of avoiding extensions for basic types.

How about this variant?

```kotlin
HttpUrl.tryParse("...")
```

File: subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt:96]: don't think so;
it looks like a function that does actual combining;
`onSuccess` on other hand mimics side effect function name

File: subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/HttpRemoteStorage.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/HttpRemoteStorage.kt]: done

File: subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt:96]: Ok

File: subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/ContentType.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.avito.filestorage
+
+import okhttp3.MediaType
+import okhttp3.MediaType.Companion.toMediaType
+
+public enum class ContentType {
+
+    PNG, MP4, TXT, HTML;
+
+    internal fun toMediaType(): MediaType = when (this) {
```
Comment (eugene-krivobokov) [inline subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/ContentType.kt:10]: We could store it as a property inside enum

File: subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/RemoteStorageFactory.kt [hunk 6]
```diff
@@ -1,26 +1,59 @@
...
+public object RemoteStorageFactory {
 
-    fun create(
-        endpoint: String,
+    public fun create(
+        endpoint: HttpUrl,
         httpClientProvider: HttpClientProvider,
-        loggerFactory: LoggerFactory,
-        timeProvider: TimeProvider,
+        isAndroidRuntime: Boolean,
```
Comment (sergioboi) [inline subprojects/common/file-storage/src/main/kotlin/com/avito/filestorage/RemoteStorageFactory.kt:16]: prefer to see two different functions comparing to flag
`createForJvm`
`createForAndroidRuntime`

File: subprojects/common/okhttp/src/main/kotlin/com/avito/http/HttpUrl.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package com.avito.http
+
+import com.avito.android.Result
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+
+public fun String.toHttpUrlResult(): Result<HttpUrl> {
+    return try {
+        Result.Success(toHttpUrl())
+    } catch (e: IllegalArgumentException) {
```
Comment (sergioboi) [inline subprojects/common/okhttp/src/main/kotlin/com/avito/http/HttpUrl.kt:10]: Why we catch only `IllegalArgumentException`?

File: subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt (hunk 3 above)
Comment (sergioboi) [inline subprojects/common/result/src/main/kotlin/com/avito/android/Result.kt:96]: This type of function usually called `zip`

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/AvitoFileStorageUploader.kt [hunk 8]
```diff
@@ -1,48 +1,30 @@
...
-                Entry.File.Type.video ->
-                    throw IllegalArgumentException("Unsupported type $type ; direct file upload should be used")
-            }
-        }.flatMap { request ->
-            when (val result = remoteStorage.upload(request, comment = "").get()) {
-                is RemoteStorage.Result.Success -> Result.Success(remoteStorage.fullUrl(result).toHttpUrl())
-                is RemoteStorage.Result.Error -> Result.Failure(result.t)
-            }
-        }
+        return remoteStorage.upload(content, type.toContentType()).get()
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/AvitoFileStorageUploader.kt:15]: So we block the whole thread while the file is uploading
From what thread do we call `upload`?

**Pull request merged**